extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java

```diff
@@ -527,102 +728,107 @@ public void run()
-              new KafkaDataSourceMetadata(finalPartitions)
-          );
-        } else {
-          action = new SegmentTransactionalInsertAction(segments, null, null);
-        }
-
-        log.info("Publishing with isTransaction[%s].", ioConfig.isUseTransaction());
-
-        return toolbox.getTaskActionClient().submit(action).isSuccess();
-      };
-
-      // Supervised kafka tasks are killed by KafkaSupervisor if they are stuck during publishing segments or waiting
-      // for hand off. See KafkaSupervisorIOConfig.completionTimeout.
-      final SegmentsAndMetadata published = driver.publish(
-          publisher,
-          committerSupplier.get(),
-          sequenceNames.values()
-      ).get();
+      // add Sentinel SequenceMetadata to indicate end of all sequences
+      publishQueue.add(SequenceMetadata.getSentinelSequenceMetadata());
+      waitForPublishes.await();
+      checkAndMaybeThrowException();
 
-      final Future<SegmentsAndMetadata> handoffFuture = driver.registerHandoff(published);
-      final SegmentsAndMetadata handedOff;
+      final List<SegmentsAndMetadata> handedOffList;
       if (tuningConfig.getHandoffConditionTimeout() == 0) {
-        handedOff = handoffFuture.get();
+        handedOffList = Futures.allAsList(handOffWaitList).get();
       } else {
-        handedOff = handoffFuture.get(tuningConfig.getHandoffConditionTimeout(), TimeUnit.MILLISECONDS);
+        handedOffList = Futures.allAsList(handOffWaitList)
+                               .get(tuningConfig.getHandoffConditionTimeout(), TimeUnit.MILLISECONDS);
       }
 
-      if (handedOff == null) {
-        throw new ISE("Transaction failure publishing segments, aborting");
-      } else {
-        log.info(
-            "Published segments[%s] with metadata[%s].",
-            Joiner.on(", ").join(
-                Iterables.transform(
-                    handedOff.getSegments(),
-                    new Function<DataSegment, String>()
-                    {
-                      @Override
-                      public String apply(DataSegment input)
-                      {
-                        return input.getIdentifier();
-                      }
-                    }
-                )
-            ),
-            handedOff.getCommitMetadata()
-        );
+      for (SegmentsAndMetadata handedOff : handedOffList) {
+        if (handedOff == null) {
+          log.warn("Handoff failed for segments [%s]", handedOff.getSegments());
+        } else {
+          log.info(
+              "Handoff completed for segments[%s] with metadata[%s].",
+              Joiner.on(", ").join(
+                  handedOff.getSegments().stream().map(DataSegment::getIdentifier).collect(Collectors.toList())
+              ),
+              handedOff.getCommitMetadata()
+          );
+        }
       }
     }
     catch (InterruptedException | RejectedExecutionException e) {
+      publishExecService.shutdownNow();
+      appenderator.closeNow();
```

Comment:
This seems redundant with the `closeNow()` call in the finally block

Reply:
We want to use `closeNow` in two cases - first when middle managers asks the task to stop, which is this case where an InterruptedException would be thrown. The second case is when some exception is thrown during processing and we want to fail the task which what `closeNow` in finally captures. So, I could not find any easy way to capture the first case in finally block, therefore an explicit call here. Does this make sense ?

Reply:
Ah okay. One way you could do it is to just set `throwableAtomicReference` in the InterruptedException catch, but this is fine.